pkg/services/ngalert/state/state.go

```diff
@@ -357,16 +362,42 @@ func resultNoData(state *State, rule *models.AlertRule, result eval.Result, logg
 			state.SetNoData("", result.EvaluatedAt, nextEndsAt)
 		}
 	case models.OK:
-		logger.Debug("Execution no data state is Normal", "handler", "resultNormal", "previous_handler", "resultNoData")
-		resultNormal(state, rule, result, logger)
-		state.StateReason = models.StateReasonNoData
+		logger.Debug("Execution no data state is Normal", "handler", "resultNormal", "previous_handler", handlerStr)
+		resultNormal(state, rule, result, logger, models.StateReasonNoData)
+	case models.KeepLast:
+		resultKeepLast(state, rule, result, logger, handlerStr)
 	default:
 		err := fmt.Errorf("unsupported no data state: %s", rule.NoDataState)
 		state.SetError(err, state.StartsAt, nextEndsTime(rule.IntervalSeconds, result.EvaluatedAt))
 		state.Annotations["Error"] = err.Error()
 	}
 }
 
+func resultKeepLast(state *State, rule *models.AlertRule, result eval.Result, logger log.Logger, previousHandler string) {
```

Comment:
It looks like you use `previousHandler` for just logging purposes. I would recommend a little bit more elegant approach
```
logg := logger.New("previous_handler", "noData")
resultKeepLast(...,logg,...)
```